File: app/src/main/java/at/bitfire/davdroid/log/PlainTextFormatter.java
/*
 * Copyright © 2013 – 2016 Ricki Hirner (bitfire web engineering).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 */

package at.bitfire.davdroid.log;

import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.exception.ExceptionUtils;
import org.apache.commons.lang3.time.DateFormatUtils;

import java.util.logging.Formatter;
import java.util.logging.LogRecord;

public class PlainTextFormatter extends Formatter {
    public final static PlainTextFormatter
            LOGCAT = new PlainTextFormatter(true),
            DEFAULT = new PlainTextFormatter(false);

    private final boolean logcat;

    private PlainTextFormatter(boolean onLogcat) {
        this.logcat = onLogcat;
    }

    @Override
    @SuppressWarnings("ThrowableResultOfMethodCallIgnored")
    public String format(LogRecord r) {
        StringBuilder builder = new StringBuilder();

        if (!logcat)
            builder .append(DateFormatUtils.format(r.getMillis(), "yyyy-MM-dd HH:mm:ss"))
                    .append(" ").append(r.getThreadID()).append(" ");

        builder.append(String.format("[%s] %s", shortClassName(r.getSourceClassName()), r.getMessage()));

        if (r.getThrown() != null)
            builder .append("\nEXCEPTION ")
                    .append(ExceptionUtils.getStackTrace(r.getThrown()));

        if (r.getParameters() != null) {
            int idx = 1;
            for (Object param : r.getParameters())
                builder.append("\n\tPARAMETER #").append(idx++).append(" = ").append(param);
        }

        if (!logcat)
            builder.append("\n");

        return builder.toString();
    }

    private String shortClassName(String className) {
        String s = StringUtils.replace(className, "at.bitfire.davdroid.", "");
        return StringUtils.replace(s, "at.bitfire.", "");
    }

}

File: app/src/main/kotlin/at/bitfire/davdroid/log/PlainTextFormatter.kt
/*
 * Copyright © All Contributors. See LICENSE and AUTHORS in the root directory for details.
 */

package at.bitfire.davdroid.log

import java.io.PrintWriter
import java.io.StringWriter
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.logging.Formatter
import java.util.logging.LogRecord

class PlainTextFormatter(
    private val withTime: Boolean,
    private val withSource: Boolean,
    private val padSource: Int = 30,
    private val withException: Boolean,
    private val lineSeparator: String?
): Formatter() {

    companion object {

        /**
         * Formatter intended for logcat output.
         */
        val LOGCAT = PlainTextFormatter(
            withTime = false,
            withSource = false,
            withException = false,
            lineSeparator = null
        )

        /**
         * Formatter intended for file output.
         */
        val DEFAULT = PlainTextFormatter(
            withTime = true,
            withSource = true,
            withException = true,
            lineSeparator = System.lineSeparator()
        )

        fun shortClassName(className: String) = className
            .replace(Regex("^at\\.bitfire\\.(dav|cert4an|dav4an|ical4an|vcard4an)droid\\."), ".")
            .replace(Regex("\\$.*$"), "")

        private fun stackTrace(ex: Throwable): String {
            val writer = StringWriter()
            ex.printStackTrace(PrintWriter(writer))
            return writer.toString()
        }

    }

    private val timeFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.ROOT)


    override fun format(r: LogRecord): String {
        val builder = StringBuilder()

        if (withTime)
            builder .append(timeFormat.format(Date(r.millis)))
                    .append(" ").append(r.threadID).append(" ")

        if (withSource && r.sourceClassName != null) {
            val className = shortClassName(r.sourceClassName)
            if (className != r.loggerName) {
                val classNameColumn = "[$className] ".padEnd(padSource)
                builder.append(classNameColumn)
            }
        }

        builder.append(r.message)

        if (withException && r.thrown != null) {
            val indentedStackTrace = stackTrace(r.thrown)
                .replace("\n", "\n\t")
                .removeSuffix("\t")
            builder.append("\n\tEXCEPTION ").append(indentedStackTrace)
        }

        r.parameters?.let {
            for ((idx, param) in it.withIndex())
                builder.append("\n\tPARAMETER #").append(idx).append(" = ").append(param)
        }

        if (lineSeparator != null)
            builder.append(lineSeparator)

        return builder.toString()
    }

}